### PR TITLE
fix: Fix repeated 403 due to app namespace being undefined (#20699)

### LINKED ISSUE #20699: appNamespace=undefined
Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

After upgrading from `2.12.0` to `2.13.0` we found that resources show progress status when it's already healthy.
This means UI can't get actual status from Argo CD server.

If we hit button `Refresh` or `Hard refresh` - not helps.

If we do regular refresh of window in browser - only after this it helps to get actual state.

I see error in developer console of browser
image

For some reason it's doing calls to:
```
appNamespace=undefined
```

**To Reproduce**

Upgrade to `2.13.0` version

**Expected behavior**

No error

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.
```

**Logs**

```
Paste any relevant application logs here.
```

### PR DESCRIPTION
**Pull request by andrii-korotkov-verkada**
Fixes #20699

Constructor may not get called every time the application changes, so previous this.appNamespace could be stale. But the update to use `this.props.match.params.appnamespace` could also fail if it's undefined. As a fix, create and use a helper function `getAppNamespace` which has a special case handling for undefined.

Also, use a namespaced endpoint when namespace is not undefined.

It needs to be cherry-picked to v2.11-2.13.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by andrii-korotkov-verkada** [comment]
> ### ✅ Preview Environment deployed on [Bunnyshell](https://www.bunnyshell.com)
> Component 	Endpoints
> argocd 	https://argocd-m3vpi2.bunnyenv.com/
> argocd-ttyd 	https://argocd-web-cli-m3vpi2.bunnyenv.com/
> 
> See: [Environment Details](https://environments.bunnyshell.com/b/environment/A5EWl01xj3) | [Pipeline Logs](https://environments.bunnyshell.com/b/pipeline/jgWvLpXbMy)
> 
> Available commands (reply to this comment):
> 
>     * 🔴 `/bns:stop` to stop the environment
> 
>     * 🚀 `/bns:deploy` to redeploy the environment
> 
>     * ❌ `/bns:delete` to remove the environment

/bns:deploy

**Comment by crenshaw-dev** [comment]
> ### ✅ Preview Environment deployed on [Bunnyshell](https://www.bunnyshell.com)
> Component	Endpoints
> argocd	https://argocd-m3vpi2.bunnyenv.com/
> argocd-ttyd	https://argocd-web-cli-m3vpi2.bunnyenv.com/
> See: [Environment Details](https://environments.bunnyshell.com/b/environment/A5EWl01xj3) | [Pipeline Logs](https://environments.bunnyshell.com/b/pipeline/jgWvLpXbMy)
> 
> Available commands (reply to this comment):
> 
> * 🔴 `/bns:stop` to stop the environment
> * 🚀 `/bns:deploy` to redeploy the environment
> * ❌ `/bns:delete` to remove the environment

/bns:deploy

**Comment by andrii-korotkov-verkada** [comment]
I've checked Bunnyshell and didn't get any 403 while inspecting network stuff. The resource tree endpoint returns 200.

**Comment by pasha-codefresh** [comment]
Asked UI team to help with review. Thanks Andrii for fixing it, once we cherry-pick it , i will create release

File: ui/src/app/applications/components/applications-list/applications-table.tsx [hunk 1]
```diff
@@ -59,7 +66,16 @@ export const ApplicationsTable = (props: {
                 applications-list__entry applications-list__entry--health-${app.status.health.status} ${selectedApp === i ? 'applications-tiles__selected' : ''}`}>
                                         <div
                                             className={`row applications-list__table-row`}
-                                            onClick={e => ctx.navigation.goto(`applications/${app.metadata.namespace}/${app.metadata.name}`, {}, {event: e})}>
+                                            // Use not namespaced endpoint if namespace is undefined
+                                            onClick={e =>
+                                                ctx.navigation.goto(
+                                                    typeof app.metadata.namespace !== 'undefined'
+                                                        ? `/applications/${app.metadata.namespace}/${app.metadata.name}`
```
Comment (oleksandr-codefresh) [inline ui/src/app/applications/components/applications-list/applications-table.tsx:73]: i see these links in two places, so it's better to move this logic into dedicated function

File: ui/src/app/applications/components/applications-list/applications-tiles.tsx [hunk 2]
```diff
@@ -66,7 +66,14 @@ export const ApplicationTiles = ({applications, syncApplication, refreshApplicat
         keys: Key.ENTER,
         action: () => {
             if (selectedApp > -1) {
-                ctxh.navigation.goto(`/applications/${applications[selectedApp].metadata.name}`);
+                let targetUrl;
+                // Use not namespaced endpoint if namespace is undefined
+                if (typeof applications[selectedApp].metadata.namespace !== 'undefined') {
+                    targetUrl = `/applications/${applications[selectedApp].metadata.namespace}/${applications[selectedApp].metadata.name}`;
```
Comment (oleksandr-codefresh) [inline ui/src/app/applications/components/applications-list/applications-tiles.tsx:72]: re-use dedicated function for that

File: ui/src/app/applications/components/applications-list/applications-tiles.tsx [hunk 3]
```diff
@@ -118,8 +125,15 @@ export const ApplicationTiles = ({applications, syncApplication, refreshApplicat
                                             }`}>
                                             <div
                                                 className='row applications-tiles__wrapper'
+                                                // Use not namespaced endpoint if namespace is undefined
                                                 onClick={e =>
-                                                    ctx.navigation.goto(`applications/${app.metadata.namespace}/${app.metadata.name}`, {view: pref.appDetails.view}, {event: e})
+                                                    ctx.navigation.goto(
+                                                        typeof app.metadata.namespace !== 'undefined'
+                                                            ? `/applications/${app.metadata.namespace}/${app.metadata.name}`
```
Comment (oleksandr-codefresh) [inline ui/src/app/applications/components/applications-list/applications-tiles.tsx:132]: re-use dedicated function for that

File: ui/src/app/applications/components/application-details/application-details-app-dropdown.tsx [hunk 4]
```diff
@@ -42,7 +42,16 @@ export const ApplicationsDetailsAppDropdown = (props: {appName: string}) => {
                                 })
                                 .slice(0, 100) // take top 100 results after filtering to avoid performance issues
                                 .map(app => (
-                                    <li key={app.metadata.name} onClick={() => ctx.navigation.goto(`/applications/${app.metadata.namespace}/${app.metadata.name}`)}>
+                                    <li
+                                        key={app.metadata.name}
+                                        onClick={() =>
+                                            ctx.navigation.goto(
+                                                // Use not namespaced endpoint if namespace is undefined
+                                                typeof app.metadata.namespace !== 'undefined'
```
Comment (oleksandr-codefresh) [inline ui/src/app/applications/components/application-details/application-details-app-dropdown.tsx:50]: use `isUndefined` from lodash

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 5]
```diff
@@ -116,6 +110,14 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
         services.extensions.removeEventListener('topBar', this.onExtensionsUpdate);
     }
 
+    private getAppNamespace() {
+        if (typeof this.props.match.params.appnamespace === 'undefined') {
+            return '';
+        } else {
```
Comment (oleksandr-codefresh) [inline ui/src/app/applications/components/application-details/application-details.tsx:116]: no need to wrap it it `else {}`
just return

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 6]
```diff
@@ -116,6 +110,14 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
         services.extensions.removeEventListener('topBar', this.onExtensionsUpdate);
     }
 
+    private getAppNamespace() {
+        if (typeof this.props.match.params.appnamespace === 'undefined') {
```
Comment (oleksandr-codefresh) [inline ui/src/app/applications/components/application-details/application-details.tsx:114]: use isUndefined from lodash in all places

**Comment by andrii-korotkov-verkada** [comment]
@oleksandr-codefresh, @plakyda-codefresh, I've addressed the feedback. Thanks!

**Comment by andrii-korotkov-verkada** [comment]
Hm, jest is unhappy with using lodash-es.
Also, it uses
```
function isUndefined(value) {
  return value === undefined;
}
```
which may not work in older browsers. I'll switch back to using a custom implementation.

**Comment by oleksandr-codefresh** [comment]
@andrii-korotkov-verkada still, create dedicated utils function `isUndefined` to avoid duplication

**Comment by andrii-korotkov-verkada** [comment]
But it's so simple, that it'd be more code to import a helper function and use it comparing to just check inline.

File: ui/src/app/applications/components/utils.tsx [hunk 7]
```diff
@@ -1471,3 +1471,10 @@ export const userMsgsList: {[key: string]: string} = {
     groupNodes: `Since the number of pods has surpassed the threshold pod count of 15, you will now be switched to the group node view.
                  If you prefer the tree view, you can simply click on the Group Nodes toolbar button to deselect the current view.`
 };
+
+export function getAppUrl(app: appModels.Application) {
```
Comment (plakyda-codefresh) [inline ui/src/app/applications/components/utils.tsx:1475]:
```suggestion
export function getAppUrl(app: appModels.Application): string {
```

Comment (andrii-korotkov-verkada) [reply ui/src/app/applications/components/utils.tsx]: Updated.

Comment (plakyda-codefresh) [reply ui/src/app/applications/components/utils.tsx]: @andrii-korotkov-verkada, could you please add a return type?

Comment (andrii-korotkov-verkada) [reply ui/src/app/applications/components/utils.tsx]: It's added in a new version of the PR.

**Comment by pasha-codefresh** [comment]
@andrii-korotkov-verkada let me know once it is ready to review, i will run it locally and see if it works

**Comment by andrii-korotkov-verkada** [comment]
@pasha-codefresh, it should be ready to review.

**Comment by pasha-codefresh** [comment]
> ### ✅ Preview Environment deployed on [Bunnyshell](https://www.bunnyshell.com)
> Component	Endpoints
> argocd	https://argocd-m3vpi2.bunnyenv.com/
> argocd-ttyd	https://argocd-web-cli-m3vpi2.bunnyenv.com/
> See: [Environment Details](https://environments.bunnyshell.com/b/environment/A5EWl01xj3) | [Pipeline Logs](https://environments.bunnyshell.com/b/pipeline/vE6Al2eJW9)
> 
> Available commands (reply to this comment):
> 
> * 🔴 `/bns:stop` to stop the environment
> * 🚀 `/bns:deploy` to redeploy the environment
> * ❌ `/bns:delete` to remove the environment

/bns:deploy

**Comment by pasha-codefresh** [comment]
@andrii-korotkov-verkada i am fine with this fix, but any reason why appNamespace should be undefined when you use search bar? i think we also should fix this in order to make it consistent. I can take a look into it

**Review by pasha-codefresh — APPROVED** [review]

**Comment by andrii-korotkov-verkada** [comment]
@pasha-codefresh, I haven't figured that out.

**Comment by pasha-codefresh** [comment]
I will check it

**Pull request merged**